topostats/tracing/dnatracing.py

```diff
@@ -79,12 +79,18 @@ def __init__(
 
     def trace_dna(self):
         """Perform DNA tracing."""
-        self.get_numpy_arrays()
-        self.gaussian_filter()
+        self.grains = self.binary_img_to_dict(self.grains_orig)
+        # What purpose does binary dilation serve here? Name suggests some sort of smoothing around the edges and
+        # the resulting array is used as a mask during the skeletonising process.
+        self.dilate_grains()
+        for grain_num, grain in self.grains.items():
+            skeleton = getSkeleton(self.gauss_image, grain).get_skeleton(self.skeletonisation_method)
+            pruned_skeleton = pruneSkeleton(self.gauss_image, skeleton).prune_skeleton(self.pruning_method)
+            self.skeleton_dict[grain_num] = pruned_skeleton #if pruned_skeleton[pruned_skeleton==1].size < 10 else pass
         self.get_disordered_trace()
```

Comment:
This and the subsequent calls should be refactored to work with a single molecule/grain and moved into the `for:` loop on line 86.

Reply:
I'm actually hoping to remove the dependancy for the single molecule dictionary and skeletonise the whole grain image at once, hence why I've left this like this. (all of the skeletonisation funcs can work do skeletonise a whole image of multiple skeletons).

This is due to the dilate function which means that grains may overlap so it is necessary to have each grain as a separate image and iterate over the single grain images. However, the dilation may close interior holes which can make the skeletons not follow their intended path.

What do you think @ns-rse?

Reply:
Not sure, that confuses me as you first say a single image with multiple skeletons can be processed but then that dilation can cause grains to overlap and so they then need iterating over individually.

`Grains` returns individual grains (as a dictionary from memory). These grains are at this stage the single unit of analysis for tracing (each is traced on its own) and any subsequent analysis rather than the image as a whole. Thus having `DNATracing` class handle a single grain/molecule seems most appropriate. It could then be inherited by another class (`TraceAll` for want of a better name) that handles all grains from an image and basically just loops over the dictionary returned by `Grains`.

Reply:
Sorry for the confusion. Yep, currently all molecules are skeletonised, pruned and traced separately, but I've made it so all the skeletonisation and pruning functions can work on the whole image, where I think this works best and quickest.

I could then move the subsequent functions within the for loop to do the tracing if you'd like? But for now due to the dilation, I have to keep the skeletonisation and pruning in the for loop.